rpcs3/Emu/Cell/Modules/cellRtc.cpp

```diff
@@ -620,124 +644,98 @@ error_code cellRtcParseRfc3339(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDate
+	date_time->microsecond = 0;
+
+	u32 pos = 19;
+	if (pszDateTime[pos] == '.')
 	{
-		date_time->microsecond = 0;
-		pc_var_5                 = 0x14;
-		if ((pszDateTime[0x14] - 0x30U) < 10)
+		u32 mul = 100000;
+
+		for (char c = pszDateTime[++pos]; std::isdigit(c); c = pszDateTime[++pos])
 		{
-			u_var_7                  = 10000;
-			date_time->microsecond = (pszDateTime[pc_var_5] + -0x30) * 100000;
-			while (true)
-			{
-				pc_var_5 = pc_var_5 + 1;
-				if (9 < (pszDateTime[pc_var_5] - 0x30U))
-					break;
-				date_time->microsecond += u_var_7 * (pszDateTime[pc_var_5] + -0x30);
-				if (u_var_7 == 100000)
-				{
-					u_var_7 = 10000;
-				}
-				else
-				{
-					if (u_var_7 == 10000)
-					{
-						u_var_7 = 1000;
-					}
-					else
-					{
-						if (u_var_7 == 1000)
-						{
-							u_var_7 = 100;
-						}
-						else
-						{
-							if (u_var_7 == 100)
-							{
-								u_var_7 = 10;
-							}
-							else
-							{
-								u_var_7 = ((u_var_7 ^ 10) - 1) >> 0x1f;
-							}
-						}
-					}
-				}
-			}
+			date_time->microsecond += digit(c) * mul;
+			mul /= 10;
 		}
 	}
-	else
-	{
-		date_time->microsecond = 0;
-	}
 
-	char_1 = pszDateTime[pc_var_5];
-	if ((char_1 == 'Z') || (char_1 == 'z'))
+	const char sign = pszDateTime[pos];
+
+	if (sign != 'Z' && sign != 'z' && sign != '+' && sign != '-')
 	{
-		l_add = 0;
+		return CELL_RTC_ERROR_BAD_PARSE;
 	}
-	else
+
+	s64 minutes_to_add = 0;
+
+	// Time offset: 1995-12-03T13:23:00.00+02:30
+	if (sign == '+' || sign == '-')
 	{
-		if ((char_1 != '+') && (char_1 != '-'))
+		if (!std::isdigit(pszDateTime[pos + 1]) ||
+			!std::isdigit(pszDateTime[pos + 2]) ||
+			pszDateTime[pos + 3] != ':' ||
+			!std::isdigit(pszDateTime[pos + 4]) ||
+			!std::isdigit(pszDateTime[pos + 5]))
 		{
 			return CELL_RTC_ERROR_BAD_PARSE;
 		}
-		char_2 = pszDateTime[pc_var_5 + 2];
-		if (((9 < (pszDateTime[pc_var_5 + 1] - 0x30U)) || (char_2 < '0')) || (i_var_6 = pszDateTime[pc_var_5 + 1] * 10 + char_2 + -0x210, '9' < char_2))
-		{
-			i_var_6 = -1;
-		}
-		char_2 = pszDateTime[pc_var_5 + 5];
-		if (((9 < (pszDateTime[pc_var_5 + 4] - 0x30U)) || (char_2 < '0')) || (i_var_8 = pszDateTime[pc_var_5 + 4] * 10 + char_2 + -0x210, '9' < char_2))
-		{
-			i_var_8 = -1;
-		}
-		if (((i_var_6 < 0) || (pszDateTime[pc_var_5 + 3] != ':')) || (i_var_8 < 0))
-		{
-			return CELL_RTC_ERROR_BAD_PARSE;
-		}
-		i_var_8 += i_var_6 * 0x3c;
-		l_add = -i_var_8;
-		if (char_1 == '-')
+
+		// Time offset (hours): 1995-12-03T13:23:00.00+XX:30
+		const s32 hours = digit(pszDateTime[pos + 1]) * 10 + digit(pszDateTime[pos + 2]);
+
+		// Time offset (minutes): 1995-12-03T13:23:00.00+02:XX
+		const s32 minutes = digit(pszDateTime[pos + 4]) * 10 + digit(pszDateTime[pos + 5]);
+
+		minutes_to_add = hours * 60 + minutes;
+
+		if (sign == '-')
 		{
-			l_add = i_var_8;
+			minutes_to_add = -minutes_to_add;
 		}
 	}
 
 	cellRtcGetTick(date_time, pUtc);
-	cellRtcTickAddMinutes(pUtc, pUtc, l_add);
+	cellRtcTickAddMinutes(pUtc, pUtc, minutes_to_add);
```

Comment:
```suggestion
	cellRtcTickAddMinutes(pUtc, pUtc, -minutes_to_add);
```
We need to _remove_ the time zone adjustment, the preceding string is already in local time and the function always returns UTC.

Reply:
The sign is already applied in the previous scope though

Reply:
Wait, am I stupid? You need to invert the sign, so for example if the string is "+01:00" then you need to subtract 60 minutes from the final tick and if it's "-01:00" you need to add 60 minutes.